### PR TITLE
Initial update to ListView functionality

### PR DESCRIPTION
**Pull request by gethinwebster**
Re-build ListView from the core react-native component, to get better
feature parity. See #260

Main changes currently:

- Incremental rendering (initialListSize/pageSize)
- Notification when at end (onEndReached, onEndReachedThreshold)

### CODE REVIEW

**Comment by gethinwebster** [comment]
Would be good to get your thoughts on this - a very rough attempt at it so far, will be testing it out more thoroughly over the next week or so so there may be some changes, but if you have any initial feedback that'd be great to hear.

**Comment by necolas** [comment]
Hey, sorry I've been swamped. This looks decent; I'll give it a proper look but would be happy to merge and let you keep iterating.

File: src/components/ListView/index.js [hunk 1]
```diff
@@ -26,74 +33,360 @@ class ListView extends Component {
...
+
+  componentDidUpdate() {
+    requestAnimationFrame(() => {
+      this._measureAndUpdateScrollProps();
+    });
   }
 
   getScrollResponder() {
-    return this._scrollViewRef && this._scrollViewRef.getScrollResponder();
+    return this._ScrollComponentRef && this._ScrollComponentRef.getScrollResponder();
```
Comment (necolas) [inline src/components/ListView/index.js]: Please revert these changes

File: src/components/ListView/index.js [hunk 2]
```diff
@@ -26,74 +33,360 @@ class ListView extends Component {
...
-    return this._scrollViewRef && this._scrollViewRef.setNativeProps(props);
+    return this._ScrollComponentRef && this._ScrollComponentRef.setNativeProps(props);
   }
 
   _onRowHighlighted(sectionId, rowId) {
     this.setState({ highlightedRow: { sectionId, rowId } });
   }
 
   render() {
+    const bodyComponents = [];
```
Comment (necolas) [inline src/components/ListView/index.js]: rename to children

File: src/components/ListView/index.js [hunk 3]
```diff
@@ -26,74 +33,360 @@ class ListView extends Component {
...
-        const key = `r_${sectionId}_${rowId}`;
-        const child = <View key={key}>{renderRow(row, sectionId, rowId, this.onRowHighlighted)}</View>;
-        children.push(child);
+      if (this.props.renderSectionHeader) {
+        const shouldUpdateHeader = rowCount >= this._prevRenderedRowsCount &&
+          dataSource.sectionHeaderShouldUpdate(sectionIdx);
+        bodyComponents.push(
+          <StaticRenderer
+            key={`s_${sectionID}`}
+            // eslint-disable-next-line react/jsx-no-bind
```
Comment (necolas) [inline src/components/ListView/index.js]: The reason bind is disallowed is for performance reasons. This can be implemented as an instance method

File: src/components/ListView/index.js [hunk 4]
```diff
@@ -26,74 +33,360 @@ class ListView extends Component {
...
+    }
+
+    const {
+      renderScrollComponent,
+      ...props
+    } = this.props;
+    if (!props.scrollEventThrottle) {
+      props.scrollEventThrottle = DEFAULT_SCROLL_CALLBACK_THROTTLE;
+    }
+    if (props.removeClippedSubviews === undefined) {
```
Comment (necolas) [inline src/components/ListView/index.js]: this should be handled with a default prop value

File: src/components/ListView/index.js [hunk 5]
```diff
@@ -26,74 +33,360 @@ class ListView extends Component {
...
+  _measureAndUpdateScrollProps() {
+    const scrollComponent = this.getScrollResponder();
+    if (!scrollComponent || !scrollComponent.getInnerViewNode) {
+      return;
+    }
+
+    this._updateVisibleRows();
+  }
+
+  _onLayout(event: Object) {
```
Comment (necolas) [inline src/components/ListView/index.js]: change to `_onLayout = (event: Object) => {}` so we can remove the binds elsewhere. can also do this for other methods

File: src/components/ListView/index.js [hunk 6]
```diff
@@ -26,74 +33,360 @@ class ListView extends Component {
...
+      };
+    }, () => {
+      this._measureAndUpdateScrollProps();
+      this._prevRenderedRowsCount = this.state.curRenderedRowsCount;
+    });
+  }
+
+  _onScroll(e: Object) {
+    const isVertical = !this.props.horizontal;
+    this.scrollProperties.visibleLength = e.nativeEvent.target[
```
Comment (necolas) [inline src/components/ListView/index.js]: `master` now normalizes the `nativeEvent`; you can rebase and drop the DOM-specific adjustments here

**Comment by gethinwebster** [comment]
Changes made. The calls to `props.renderRow` and `props.renderSectionHeader` are still bind-ish, so am not certain if that's better or not - let me know if I've missed something obvious there.

**Comment by necolas** [comment]
Thanks!
